stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,80 +1,141 @@
+     * specializations for Eigen::Matrices and std::vectors (where this
+     * assumption is not true).
+     *
+     * @tparam T1 Type of x
+     * @tparam T2 Type of y
+     * @param x First argument to check
+     * @param y Second argument to check
+     */
+    template<typename T1, typename T2>
+    inline void append_array_check_size(const T1& x, const T2& y) {
+    }
+
+    /**
+     * This function checks if two Eigen::Matrices are of compatible sizes
+     * by delegating to check_matching_dims.
+     *
+     * @tparam T1 Scalar type of x
+     * @tparam T2 Scalar type of y
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x First matrix
+     * @param y Second matrix
+     */
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_check_size(const Eigen::Matrix<T1, R, C>& x,
+                     const Eigen::Matrix<T2, R, C>& y) {
+      check_matching_dims("append_array", "x", x, "y", y);
+    }
+
+    /**
+     * This function checks if two std::vectors are of compatible sizes and
+     * then recursively calls append_array_check_size to see if the first
+     * elements of each std::vector have compatible sizes.
+     *
+     * @tparam T1 Element type of x
+     * @tparam T2 Element type of y
+     * @param x First std::vector
+     * @param y Second std::vector
+     */
+    template<typename T1, typename T2>
+    inline void append_array_check_size(const std::vector<T1> &x,
+                                        const std::vector<T2> &y) {
+      check_consistent_sizes("append_array", "size of x", x, "size of y", y);
+      if (!x.empty() && !y.empty()) {
+        append_array_check_size(x[0], y[0]);
+      }
     }
 
     /**
      * Return the concatenation of two specified vectors in the order of
      *   the arguments.
      *
-     * @tparam T1 Scalar type of first vector
-     * @tparam T2 Scalar type of second vector
+     * @tparam T1 Type of first vector
+     * @tparam T2 Type of second vector
      * @param x First vector
      * @param y Second vector
      * @return A vector of x and y concatenated together (in that order)
      */
     template <typename T1, typename T2>
-    inline typename std::vector<typename return_type<T1, T2>::type>
-    append_array(const std::vector<T1>& x, const std::vector<T2>& y) {
-      std::vector<typename return_type<T1, T2>::type> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
+    inline typename append_return_type<T1, T2>::type
+    append_array(const T1& x, const T2& y) {
+      typename append_return_type<T1, T2>::type z;
+      if (!x.empty() && !y.empty()) {
+        append_array_check_size(x[0], y[0]);
+      }
+      z.resize(x.size() + y.size());
+      for (size_t i = 0; i < x.size(); i++)
+        append_array_copy(x[i], z[i]);
```

Comment:
With this structure it's clear that you need an `append_array_copy` specializion for when the two arguments are the same type---in that case, you do *not* need to do this all recursively, and it'll be *much* faster.